### ISSUE TITLE
Creating a WMS layer via PyQGIS triggers segfault

### ISSUE THREAD

**Issue by trey-stafford** [qgis/QGIS#35354]
**Describe the bug**

We have encountered a problem in attempting to create WMS layers using PyQGIS.

Attempting to create a QgsRasterLayer with the `wms` provider results in a segmentation fault.
**How to Reproduce**

```
import qgis.core as qgc

# Supply path to qgis install location
qgc.QgsApplication.setPrefixPath("/path/to/qgis/installation", True)

qgs = qgc.QgsApplication([], False)

# Load providers
qgs.initQgis()

url = "http://www.glims.org/mapservice?VERSION=1.3.0&STYLES="

print(qgc.QgsRasterLayer(url, 'GLIMS_GLACIERS', 'wms'))

```
The only output from running the above code is the following (return code 139):

```
Segmentation fault (core dumped)
```
**QGIS and OS versions**

QGIS: We have observed this problem in QGIS versions v3.10.1 and v3.12.0, both installed from conda-forge. 

OS: Ubuntu 16.04.6 LTS

**Comment by elpaso** [comment]
Is this the full script?

**Comment by gioman** [comment]
>  both installed from conda-forge.

@trey-stafford not sure if we officially support conda installations (we did not but maybe something changed), have you tested on a "normal" ubuntu installation via repositories?

**Comment by trey-stafford** [comment]
@elpaso The given code is all it takes to trigger the segfault. 

@gioman No, we have not tested against qgis installations via the ubuntu repositories. Our project relies on a number of different python dependencies, and we are using `conda` to manage those. Hoping this is not a problem inherent to the `conda-forge` installation, but definitely worth checking.

**Comment by trey-stafford** [comment]
@gioman Just tested with QGIS 3.10.3 using the `qgis/qgis:release-3_10` docker image (https://hub.docker.com/r/qgis/qgis), which resulted in the same error.

**Comment by elpaso** [comment]
You need to call initQgis before doing anything.

**Comment by trey-stafford** [comment]
@elpaso We actually did previously have the call to `initQgis` in our code, we found that it was not necessary for our purposes. Perhaps we should do so as a best practice nevertheless.

In any case, calling `initQgis` does not solve the problem at hand. I can update the example code to include that if it helps rule things out though!

**Comment by elpaso** [comment]
The provider name is `wms`, not `WMS`, the only uppercase is `WFS` (unfortunately)

**Comment by trey-stafford** [comment]
@elpaso Thanks for pointing that out. I will update the example script to reflect that.

**Comment by elpaso** [comment]
still crashing if you use `wms`?

**Comment by trey-stafford** [comment]
@elpaso yes, unfortunately still the same issue.

**Comment by MattF-NSIDC** [comment]
QGIS 3.14 release no longer segfaults for us on running the example script. We now get an invalid layer `<QgsMapLayer: 'GLIMS_GLACIERS' (Invalid)>`. That looks like a separate issue, likely on our end! Thanks for your help.

**Comment by trey-stafford** [comment]
Confirmed this also works with qgis v3.10.7.

**Comment by MattF-NSIDC** [comment]
We attempted to add the invalid layer created by this script to a project and write it and the segfault came back on the final write in the below script :frowning_face: 

```
import qgis.core as qgc

# Supply path to qgis install location
qgc.QgsApplication.setPrefixPath("/path/to/qgis/installation", True)
qgs = qgc.QgsApplication([], False)

# Load providers
qgs.initQgis()

url = 'IgnoreGetMapUrl=1&contextualWMSLegend=0&crs=EPSG:3857&dpiMode=7&featureCount=10&format=image/png&layers=GLIMS_GLACIERS&styles&tileMatrixSet=GLIMS_GLACIERS-wmsc-0&url=http://www.glims.org/mapservice?VERSION=1.3.0'
layer = qgc.QgsRasterLayer(url, 'GLIMS_GLACIERS', 'wms')

project = qgc.QgsProject.instance()
project.write('/tmp/wms_invalid.qgs')
project.addMapLayer(layer)
project.write()
```

We are aware that the layer is invalid, but as we had successfully added the same layer in the QGIS GUI, we hoped we would at some point hit a message telling us what's wrong with the layer or at least learn something new by writing the project file, examining its XML, and comparing it to the project where we successfully added this layer via the QGIS GUI.

**Comment by SurfTimeOut** [comment]
@trey-stafford  and @MattF-NSIDC , I am also experiencing an error with reading WFS into PyQGIS. I first created the project with a valid WFS layer, the extents of this layer in the GUI everything seems ok. Reading this project into PyQGIS the layer is invalid and the extents have changed. I have tried this with more than one WFS and get the same result. 

https://services.ga.gov.au/site_1/services/Geomorphology_Scale_Guide_WM/MapServer/WFSServer?request=GetCapabilities&service=WFS

I am assuming this is an error reading the geometry. 

![image](https://user-images.githubusercontent.com/22866414/97819150-e8ea4100-1cfa-11eb-9697-31bc8c56f0ad.png)

![image](https://user-images.githubusercontent.com/22866414/97819132-d07a2680-1cfa-11eb-8fa2-7c5f94ef1b60.png)

**Comment by christophfink** [comment]
Hope I’m not awakening a dead issue here, but I am experiencing something similar with a `QgsVectorLayer` with the `memory` provider. The segfault is triggered when calling `QgsProject.instance().addMapLayer()` from within a plugin (cartogram3), only on MacOS. Currently testing on different versions, but definitely happening with a fresh install of QGIS 3.20.0 on MacOS 10.15.7.
(dump by the plugin user who reported the segfault at https://github.com/austromorph/cartogram3/issues/22 )

**Comment by trey-stafford** [comment]
We've been testing our code against qgis v3.16.3 (provided by `conda-forge`) and have found that not properly cleaning up the `QgsApplication` seems to trigger segfaults. The following script uses a context manager to setup and tear down the `QgsApplication` and allows the program to complete successfully without segfaulting:

```
import os
import subprocess

import qgis.core as qgc


def _get_qgs_prefix_path() -> str:
    # The qgis prefix path is two directories above the qgis executable.
    # See:
    # https://docs.qgis.org/testing/en/docs/pyqgis_developer_cookbook/intro.html#using-pyqgis-in-standalone-scripts
    qgis_path = subprocess.run(
        ['which', 'qgis'],
        stdout=subprocess.PIPE
    ).stdout.decode('utf-8').strip('\n')
    qgis_prefix_path = os.path.abspath(os.path.join(qgis_path, '..', '..'))

    return qgis_prefix_path


def _setup_qgs_app() -> qgc.QgsApplication:
    """Set up QgsApplication.

    This function should only be called once to instantiate a
    QgsApplication. Once all pyqgis operations are complete, the QgsApplication
    MUST be cleaned up with the QgsApplication instance's `exitQgis`
    method. Failure to do so will result in a segmentation fault.
    """
    qgis_prefix_path = _get_qgs_prefix_path()

    # Boilerplate QGIS initialization code;
    # - Suppresses "Application path not intialized" errors
    # - Enables Qt support for fonts used in layer styles (labels)
    qgs = qgc.QgsApplication([], False)
    qgs.initQgis()
    qgc.QgsApplication.setPrefixPath(qgis_prefix_path, True)

    return qgs


class QgsApplicationContext:
    """Context manager for setting up and tearing down the QgsApplication.

    WARNING: This context manager should be used once and only once per program
    execution. Attempting to use this context manager more than once may cause a
    segfault.
    """

    def __enter__(self):
        self.qgs = _setup_qgs_app()

        return self.qgs

    def __exit__(self, _exc_type, _exc_value, _traceback):
        # Run `exitQgis` to properly cleanup the QgsApplication. Not doing so
        # causes a segfault on program completion.
        self.qgs.exitQgis()



if __name__ == '__main__':
    with QgsApplicationContext():
        url = 'IgnoreGetMapUrl=1&contextualWMSLegend=0&crs=EPSG:3857&dpiMode=7&featureCount=10&format=image/png&layers=GLIMS_GLACIERS&styles&tileMatrixSet=GLIMS_GLACIERS-wmsc-0&url=http://www.glims.org/mapservice?VERSION=1.3.0'
        layer = qgc.QgsRasterLayer(url, 'GLIMS_GLACIERS', 'wms')

        project = qgc.QgsProject.instance()
        project.write('/tmp/wms_invalid.qgs')
        project.addMapLayer(layer)
        project.write()

        print('Done')
```
